core/scheduler.py

```diff
@@ -16,86 +16,107 @@
-        self.results = [None]*self.num_runners
-        self.event = Event()
+        self.current_jobs = [None for _ in range(self.num_runners)]
         logging.info("Scheduler is set up!")
 
     def add_job(self, dml_job):
-        """ Add a job to the queue"""
+        """
+        Add a job to the queue.
+        """
         assert type(dml_job) is DMLJob, "Job is not of type DMLJob."
         logging.info("Scheduling job...")
         self.queue.append(dml_job)
 
-    def _runners_run_next_jobs(self):
-        """ Check each job to see if it has a job running. If not, have the runner run the
-        next job on the queue asynchronously of the others and collect the result of the job
-        that was running before (if applicable)
+    def start_cron(self, period_in_mins=None):
         """
-        for i in range(self.num_runners):
-            runner = self.runners[i]
-            if not runner.is_active():
-                if self.current_jobs[i]:
-                    self.processed.append(self.current_jobs[i].get())
-                    self.current_jobs[i] = None
-                if self.queue:
-                    self.current_jobs[i] = self.pool.apply_async(runner.run_job, (self.queue.popleft(),))
-
-    def _runners_run_next_jobs_as_event(self, period):
-        """ Trigger above method every period"""
-        self._runners_run_next_jobs()
-        if not self.event.is_set():
-            Timer(period*60, self._runners_run_next_jobs_as_event, [period]).start()
-
-    def _start_cron(self, period=None):
-        """ CRON job to run next jobs on runners, if applicable. Runs asynchronously."""
-        if not period:
-            period = self.frequency_in_mins
+        CRON job to run next jobs on runners, if applicable. Runs asynchronously.
+        """
+        if not period_in_mins:
+            period_in_mins = self.frequency_in_mins
         logging.info("Starting cron...")
-        self._runners_run_next_jobs_as_event(period)
+        self._runners_run_next_jobs_as_event(period_in_mins)
         logging.info("Cron started!")
 
+    def stop_cron(self):
+        """
+        Tell the scheduler to stop scheduling jobs.
+        """
+        logging.info("Stopping cron...")
+        self.event.set()
+        logging.info("Cron stopped!")
 
-    def reset(self):
-        """ Scheduler is a singleton, so need to get a new instance inplace """
+    def reset(self, reset_history=False):
+        """
+        Resets the scheduler.
+        """
         logging.info("Resetting scheduler...")
         self.queue = deque()
         self.processed = []
+        if reset_history:
+            self.history = []
         self.event = Event()
         logging.info("Scheduler resetted!")
 
-    def _stop_cron(self):
-        """ Tell the scheduler to stop scheduling jobs """
-        logging.info("Stopping cron...")
-        self.event.set()
-        logging.info("Cron stopped!")
+    def _runners_run_next_jobs(self):
```

Comment:
Why are these functions private? Why are they being called in the tests if they're private?

Reply:
Good catch. We had it private because we intended to run jobs only by the dataset manager's cron job, not directly by that function. But yes, tests should only be testing the public interface of the class and I can see this function being used directly by the users of the class, so I made it public 👍